swig/csharp/CMakeLists.txt

```diff
@@ -9,226 +15,624 @@ set(GDAL_CSHARP_INSTALL_DIR
-  DEPENDS ogr_csharp.dll osr_csharp.dll)
+  TARGET gdal_csharp.dll TARGET_SUBDIR gdal
+  DEPENDS osr_csharp
+          ogr_csharp
+          gdalconst_csharp
+          WRAP_DLL
+          gdal_wrap
+          SYSTEM_DEPENDS
+          OSGeo.OSR
+          OSGeo.OGR
+          OSGeo.GDAL.CONST
+          OSGeo.GDAL.Core
+          PACKAGE_NAME
+          OSGeo.GDAL)
 
 set_directory_properties(PROPERTIES ADDITIONAL_MAKE_CLEAN_FILES "gdal;ogr;osr;const")
 
 # ######################################################################################################################
 # sample commands
 
 function (gdal_build_csharp_sample)
+
+  # setup arguments
   set(_options)
   set(_oneValueArgs SOURCE OUTPUT)
   set(_multiValueArgs DEPENDS SYSTEM_DEPENDS)
   cmake_parse_arguments(_GBCS "${_options}" "${_oneValueArgs}" "${_multiValueArgs}" ${ARGN})
   set(CSC_OPTIONS)
-  set(_depends)
-  foreach (_dep IN LISTS _GBCS_DEPENDS)
-    list(APPEND CSC_OPTIONS /r:${_dep})
-    list(APPEND _depends "${CMAKE_CURRENT_BINARY_DIR}/${_dep}")
-  endforeach ()
-  foreach (_dep IN LISTS _GBCS_SYSTEM_DEPENDS)
-    list(APPEND CSC_OPTIONS /r:${_dep})
-  endforeach ()
+
+  # setup project file
+  get_filename_component(_folder ${_GBCS_OUTPUT} NAME_WE)
+  set(_GBCS_PROJ_PATH ${CMAKE_CURRENT_BINARY_DIR}/${_folder})
+  set(_GBCS_PROJ ${_GBCS_PROJ_PATH}/${_folder})
+  file(TO_NATIVE_PATH "${_GBCS_PROJ}" _GBCS_PROJ_NATIVE)
+  file(TO_NATIVE_PATH "${_GBCS_PROJ_PATH}" _GBCS_PROJ_PATH_NATIVE)
+  file(MAKE_DIRECTORY ${_GBCS_PROJ_PATH})
+
+  file(TO_NATIVE_PATH "${CMAKE_CURRENT_BINARY_DIR}/${_GBCS_OUTPUT}" _GBCS_OUTPUT_NATIVE)
+
+  # Setup dependencies
+  set(_dlls)
+  if (_GBCS_DEPENDS)
+    foreach (_dep IN LISTS _GBCS_DEPENDS)
+      file(TO_NATIVE_PATH "${CMAKE_CURRENT_BINARY_DIR}/${_dep}" _dep)
+      get_filename_component(_filename ${_dep} NAME)
+      list(APPEND _dlls ${_filename}.dll)
+      list(APPEND CSC_OPTIONS /r:${_dep}.dll)
+    endforeach ()
+  endif (_GBCS_DEPENDS)
+  if (_GBCS_SYSTEM_DEPENDS)
+    foreach (_dep IN LISTS _GBCS_SYSTEM_DEPENDS)
+      if (NOT _dep MATCHES "^OSGeo*.")
+        list(APPEND CSC_OPTIONS /r:${_dep}.dll)
+      endif ()
+    endforeach ()
+  endif (_GBCS_SYSTEM_DEPENDS)
+
   file(TO_NATIVE_PATH "${_GBCS_SOURCE}" SOURCE_NATIVE)
 
-  add_custom_command(
-    OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/${_GBCS_OUTPUT}
-    COMMAND ${CSHARP_COMPILER} ${CSC_OPTIONS} /out:${_GBCS_OUTPUT} ${SOURCE_NATIVE}
-    DEPENDS ${_depends} ${_GBCS_SOURCE})
+  # build the sample exe
+  if (DOTNET_FOUND)
+    configure_file("${CMAKE_CURRENT_SOURCE_DIR}/exe_template.csproj" "${_GBCS_PROJ}.csproj")
+    add_dotnet(
+      ${_GBCS_PROJ_NATIVE}.csproj
+      ${CSHARP_CONFIG}
+      NAME
+      OSGeo.GDAL.samples.${_folder}
+      VERSION
+      ${_VERSION_STRING}
+      OUTPUT_PATH
+      ${_GBCS_PROJ_PATH_NATIVE}
+      DEPENDS
+      csharp_binding
+      PACKAGE
+      ${_GBCS_SYSTEM_DEPENDS})
+
+  else ()
+    message(
```

Comment:
could be removed or commented to avoid too verbose log

Reply:
made into --verbose only